File: ex235.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode ancestor = root;
        
        // Make sure that p < q
        if (p.val > q.val) {
            TreeNode temp = p;
            p = q;
            q = temp;
        }
        
        
        while (ancestor != null) {
            // If two nodes p, q are on different branches of a root node, 
            // this root node is the common ancestor of them
            if ((p.val <= ancestor.val) & (q.val >= ancestor.val))
                break;
            
            // If two nodes p, q are on the same branch of a root node, 
            // move to the next root node
            if ((p.val < ancestor.val) & (q.val < ancestor.val))
                ancestor = ancestor.left;
            
            if ((p.val > ancestor.val) & (q.val > ancestor.val))
                ancestor = ancestor.right;
        }
        
        return ancestor;
    }
}

File: code/ex235.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
    TreeNode* ancestor = NULL;
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs(root, p->val, q->val);
        return ancestor;
    }
    
private:
    int dfs(TreeNode* root, int p, int q) {
        if (root == NULL)    return 0;
        int left = dfs(root->left, p, q);
        if (ancestor != NULL)    return 0;
        int right = dfs(root->right, p, q);
        if (ancestor != NULL)    return 0;
        int cur = root->val == p ? 1 : root->val == q ? 2 : 0;
        if (left + cur == 3 || left + right == 3 || right + cur == 3) {
            ancestor = root;
        }
        if (cur == 1 || left == 1 || right == 1)    return 1;
        if (cur == 2 || left == 2 || right == 2)    return 2;
        return 0;
    }
};